client/commands/v2/pysa_server.py

```diff
@@ -59,6 +65,56 @@ def __init__(
         self.pyre_arguments = pyre_arguments
         self.binary_location = binary_location
         self.server_identifier = server_identifier
+        self.pyre_connection = api_connection.PyreConnection(
+            Path(self.pyre_arguments.global_root)
+        )
+
+    def invalid_model_to_diagnostic(
+        self,
+        invalid_model: query.InvalidModel,
+    ) -> lsp.Diagnostic:
+        return lsp.Diagnostic(
+            range=lsp.Range(
+                start=lsp.Position(
+                    line=invalid_model.line - 1, character=invalid_model.column
+                ),
+                end=lsp.Position(
+                    line=invalid_model.stop_line - 1,
+                    character=invalid_model.stop_column,
+                ),
+            ),
+            message=invalid_model.full_error_message,
+            severity=lsp.DiagnosticSeverity.ERROR,
+            code=None,
+            source="Pysa",
+        )
+
+    def invalid_models_to_diagnostics(
+        self,
+        invalid_models: Sequence[query.InvalidModel],
+    ) -> Dict[Path, List[lsp.Diagnostic]]:
+        result: Dict[Path, List[lsp.Diagnostic]] = defaultdict(list)
+        for model in invalid_models:
+            if model.path is None:
+                self.log_and_show_message_to_client(
+                    f"{model.full_error_message}", lsp.MessageType.ERROR
+                )
+            else:
+                result[Path(model.path)].append(self.invalid_model_to_diagnostic(model))
+        return result
+
+    async def update_errors(self, document_path: Path) -> None:
+        # Publishing empty diagnostics to clear errors in VSCode
+        await _publish_diagnostics(self.output_channel, document_path, [])
```

Comment:
Feedback from @grievejia:

> If I understand correctly, update_errors will just query for invalid model errors in all files, regardless of whether the currently opened/saved file (i.e. document_path) is? If that's the case I think it's not enough to just clear out diagnostics in document_path. One needs to basically clean up all files in which there are pre-existing diagnostics (you'd probably need to keep track of more states in the server for that purpose).
> 
> Alternatively I wonder if it's possible to either just query invalid model errors for one file. Also, does invalid model error support incremental changes? Is it necessary to query these errors on file open, given that opening a file won't change any of these errors?

Reply:
This is a good point. Right now we don't have support for incremental or per-file model errors. We just get a list of all possible model errors. I think the simplest solution is just to keep a list of errors we've published previously, and then zero them out when we publish new errors. Thoughts on that solution @grievejia ?

Reply:
> I think the simplest solution is just to keep a list of errors we've published previously, and then zero them out when we publish new errors. 

Ah that sounds good. Does this mean that we'll only need to re-query on saved instead of opens?

Reply:
@grievejia I think we would want to query on open so that we tell the user right away if they open a file and it already has errors.

Reply:
Oh I see what you mean. 

What I had in mind was to separate the querying logic from displaying logic, i.e. we do a query as soon as the background server establishes, and updates the query whenever a file is saved. Whether the query result gets shown to the users or not, on the other hand, gets controlled by the open request (or not at all if you want to show errors in unopened files as well).

But I think it would also work if we always query on open. Functionally they should be the same.